File: dailyQuestion/2020-08/08-12/golang/solution.go
package main

import (
	"fmt"
)

func exist(board [][]byte, word string) bool {
	m := len(board)
	if m == 0 {
		return false
	}
	n := len(board[0])
	var dfs func(i, j, k int) bool
	dfs = func(i, j, k int) bool {
		if (i < 0 || i >= m) || (j < 0 || j >= n) || board[i][j] != word[k] {
			return false
		}
		if k == len(word)-1 {
			return true
		}
		tmp := board[i][j]
		board[i][j] = '/'
		res := dfs(i+1, j, k+1) || dfs(i, j+1, k+1) || dfs(i-1, j, k+1) || dfs(i, j-1, k+1)
		board[i][j] = tmp
		return res
	}

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if dfs(i, j, 0) {
				return true
			}
		}
	}
	return false
}

func main() {
	board := [][]byte{[]byte{'A', 'B', 'C', 'E'}, []byte{'S', 'F', 'C', 'S'}, []byte{'A', 'D', 'E', 'E'}}
	word := "ABCCED"
	result := exist(board, word)
	fmt.Println(result)
}

File: dailyQuestion/2020-08/08-12/python/solution.py
def exist(board: [[str]], word: str) -> bool:
    m = len(board)
    if m == 0 :
        return False
    n = len(board[0])
    def dfs(i,j,k) :
        if (i < 0 or i >= m) or (j < 0 or j >= n) or (board[i][j] != word[k]) :
            return False
        if k == len(word) - 1 :
            return True
        tmp , board[i][j] = board[i][j], '/'
        res = dfs(i+1,j,k+1) or dfs(i,j+1,k+1) or dfs(i-1,j,k+1) or dfs(i,j-1,k+1)
        board[i][j] = tmp
        return res
    for i in range(m) :
        for j in range(n) :
            if dfs(i,j,0) :
                return True
    return False

if __name__ == "__main__" :
    board = [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]]
    word = "ABCCED"
    result = exist(board,word)
    print(result)